{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/emgoss/PLP6621C/blob/main/Module1_R.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "w6haslUgv6fI"
      },
      "source": [
        "# Lab 1: Population simulations\n",
        "\n",
        "The goal of this exercise is to build an intuition for population genetic processes in populations. The R package “learnPopGen” simulates a Wright-Fisher population over time. The main parameters that we will vary are allele frequency, population size, and number of generations that the simulation is run. We can also run multiple replicate populations at the same time. After running simulations with no selection, we will compare these outcomes to what we see when there is natural selection on alleles by changing the relative fitness of alleles and genotypes.\n",
        "\n",
        "Run the commands (press \"play\" icons in code boxes) to obtain plots. Later in the exercise, I ask you to edit the command line.\n",
        "\n",
        "The first command installs the package and will take a while to run.\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "install.packages(\"learnPopGen\")"
      ],
      "metadata": {
        "id": "F5Cs9ChaL41i"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "library(learnPopGen) # This tells R to load the package now that it is installed."
      ],
      "metadata": {
        "id": "Y_VgiGgFL8_J"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Now we will run a simulation of a population with 100 individuals (Ne). These are diploids - so there are 200 copies of the gene or chromosome we are simulating. In this simulation we will track our allele of interest, allele A. Here, allele A starts at (p0=0.5), meaning it is present in half chromosomes in this population and the other half have the alternative allele (a). Each allele (A,a) and genotype (AA, Aa, aa) have equal fitness (w), meaning they contribute equally to the next generation on average. We will run the simulation for 400 generations (ngen) to see what happens to the frequency of allele A [f(A)]."
      ],
      "metadata": {
        "id": "dSUzBOsemkbt"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "tPkgvj_v1SL2"
      },
      "outputs": [],
      "source": [
        "drift.selection(p0=0.5, Ne=100, w=c(1,1,1), ngen=400, nrep=1)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "3aGuqzdh2bA6"
      },
      "source": [
        "The resulting plot will vary each time you run it. The population size stays the same though. Imagine that each generation there are many randomly combining gametes but only 100 resulting offspring."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "1sNs-w0J6xmY"
      },
      "source": [
        "You can compare multiple realizations of the simulation by increasing the number of replicates to display at one time (nrep). Now we will compare the outcome in 5 different populations."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "yV2xBoLA65P-"
      },
      "outputs": [],
      "source": [
        "drift.selection(p0=0.5, Ne=100, w=c(1,1,1), ngen=400, nrep=5)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "DzfDqrt-7GiI"
      },
      "source": [
        "When f(A)=1, allele A has reached fixation in the population, meaning all individuals have the A allele. When f(A)=0, the A allele is lost from the population and all individuals have the a allele. You may see that sometimes the A allele fixes and sometimes it is lost.\n",
        "\n",
        "\n",
        "\n",
        "---\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xweGGMbc7eQ1"
      },
      "source": [
        "Now we will vary population size (Ne) and see what happens. Run at least 20 simulations for different population sizes. You might try Ne of 10, 25, 100, and larger. You could run only one or two populations at a time (nrep) if you are having trouble seeing the dynamics or five at a time (or more if you’d like). You may need to increase or decrease the number of generations simulated to capture the dynamics."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "h3BqyCxN7GLd"
      },
      "outputs": [],
      "source": [
        "drift.selection(p0=0.5, Ne=10, w=c(1,1,1), ngen=100, nrep=5)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ZtRRQR4h8AYZ"
      },
      "outputs": [],
      "source": [
        "drift.selection(p0=0.5, Ne=25, w=c(1,1,1), ngen=200, nrep=5)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "OVj84dve8Xgi"
      },
      "outputs": [],
      "source": [
        "drift.selection(p0=0.5, Ne=500, w=c(1,1,1), ngen=1000, nrep=5)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "FB0alojJgUTM"
      },
      "source": [
        "**Question 1:** What are your observations about the allele frequencies over time at different population sizes? Comment how long it takes for fixation of A, f(A)=1, or loss of A, f(A)=0.\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "1p5vDzmSgmlt"
      },
      "source": [
        "***Your answer here (double click to type):***\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "---\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0ZBaCfLe86v3"
      },
      "source": [
        "I changed the population size and allele frequency to  1/(2N) (refer to table below), which is the frequency of a new mutation in a diploid population (i.e., 1 copy of the mutation in 2N chromosomes.)\n",
        "\n",
        "\tN    Initial frequency\n",
        "---\n",
        "\t4    0.125\n",
        "\t20   0.025\n",
        "\t50   0.01\n",
        "\t100  0.005\n",
        "\t500  0.001\n",
        "Observe the dynamics for each population size. (Note that I changed the number of reps to 20 below.)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "rmXa3fVq8whv"
      },
      "outputs": [],
      "source": [
        "drift.selection(p0=0.125, Ne=20, w=c(1,1,1), ngen=200, nrep=20)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "MNoI8thW9XAy"
      },
      "outputs": [],
      "source": [
        "drift.selection(p0=0.01, Ne=50, w=c(1,1,1), ngen=500, nrep=20)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ZRVWNV5J9sg7"
      },
      "outputs": [],
      "source": [
        "drift.selection(p0=0.005, Ne=100, w=c(1,1,1), ngen=1000, nrep=20)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "WMUZKMLK94-l"
      },
      "outputs": [],
      "source": [
        "drift.selection(p0=0.001, Ne=500, w=c(1,1,1), ngen=1000, nrep=20)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2KIcOtThhomS"
      },
      "source": [
        "**Question 2** What do you notice about the effect of population size on the dynamics of the new mutation?"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2atEQW8Hh-H9"
      },
      "source": [
        "***Your answer here:***\n",
        "\n",
        "\n",
        "\n",
        "---\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "vZvOyEOv-U6P"
      },
      "source": [
        "Now that you have some idea of the effect of drift on a new mutation, how does higher or lower fitness of this mutation affect its frequency in the population? Here, the new mutation is allele A, so it comes into the population as Aa (heterozygous) and mating between Aa and Aa creates homozygous AA genotypes. If the homozygous state of AA has slightly higher fitness than genotype aa and the heterozygote is codominant (fitness inbetween AA and aa), then we can change the fitnesses to w=c(1.2, 1.1, 1.0). Play with fitness values and see how it affects the fixation or loss of the new mutation. I gave one example below, but please try other options by editing the command."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "kSFpgxMO-hIg"
      },
      "outputs": [],
      "source": [
        "drift.selection(p0=0.005, Ne=20, w=c(1.2,1.1,1), ngen=100, nrep=50)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "SQSlsz_kaQrZ"
      },
      "source": [
        "Here the heterozygote has the same fitness as AA, meaning that A is dominant."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "hcjIwlvSaVas"
      },
      "outputs": [],
      "source": [
        "drift.selection(p0=0.005, Ne=100, w=c(1.2,1.2,1), ngen=100, nrep=20)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "SFBHVfZO-2p2"
      },
      "source": [
        "What if there is overdominance, meaning that the heterozygote has higher fitness than either homozygote?"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "k1q8GiSW-4q3"
      },
      "outputs": [],
      "source": [
        "drift.selection(p0=0.005, Ne=100, w=c(1,1.2,1), ngen=100, nrep=20)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "vHuO3r3GisID"
      },
      "source": [
        "**Question 3:** Discuss how selection on a new mutation interacts with population size. What are the consequences of selection for A when:\n",
        "\n",
        "1.   A is a codominant allele (fitness of AA>Aa>aa)?\n",
        "2.   A is the dominant allele (AA=Aa>aa)?\n",
        "3.   there is heterozygote advantage (Aa > AA and aa)?\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-mm3v4aOkbhv"
      },
      "source": [
        "***Your answer:***\n",
        "\n",
        "\n",
        "\n",
        "---\n",
        "\n"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyPzrjU8C5wk8l8Ui3rzwdgY",
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "R",
      "name": "ir"
    },
    "language_info": {
      "name": "R"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}