{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyNBXp06zTnM2NiFUUcgnJsK",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/emgoss/PLP6621C/blob/main/Module3.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Module 3 Assignment: Model Selection\n",
        "\n",
        "We will use IQTREE to select the best model for our sequence alignment.\n",
        "\n",
        "IQTREE is a flexible phylogenetic program with pretty good documentation.\n",
        "\n",
        "\n",
        "If you'd like to learn more about IQTREE, see the tutorials here:\n",
        "[Beginner’s tutorial](https://iqtree.github.io/doc/Tutorial)\n",
        "\n",
        "\n",
        "---\n",
        "\n"
      ],
      "metadata": {
        "id": "i4W8OUMVGYJZ"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**1. Sign up for colab Pro for college students\n",
        "https://colab.research.google.com/signup. Click \"No cost for students and educators\"**"
      ],
      "metadata": {
        "id": "hnWxDl455Ca-"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**2. Install IQTREE.**\n",
        "\n",
        "I got help installing and running the IQTREE program from the built in Colab Gemini chatbot. This is a great feature of Colab if you're not comfortable coding or using this interface.\n",
        "\n",
        "The following text boxes are from Gemini."
      ],
      "metadata": {
        "id": "KcUeEjQYC1LT"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "heOs72QTvGL0"
      },
      "outputs": [],
      "source": [
        "!apt-get install iqtree"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "26f52711"
      },
      "source": [
        "The below command will list all files installed by the `iqtree` Debian package. Look for a file named `iqtree` in the output, typically within a `bin` directory (e.g., `/usr/bin/iqtree`). Once we find the correct path, we can use it to run the tool."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "1cd900bf"
      },
      "source": [
        "!dpkg -L iqtree"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "40e2b270"
      },
      "source": [
        "The `which` command should output the full path to the `iqtree` executable if it's in the system's PATH. If it doesn't output anything, it means `iqtree` might not be in a standard location or not fully installed correctly. If a path is returned, we can then use that path in the command."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "26f30d3d"
      },
      "source": [
        "!which iqtree2"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**We will determine the best model for the DNA sequence alignment HPPD.fasta using IQTREE's ModelFinder (Kalyaanamoorthy et al 2017) and then compute a phylogenetic tree.**\n",
        "\n",
        "The file HPPD.fasta contains sequences from a single locus amplified from isolates of the wheat pathogen *Zymoseptoria tritici* from the Middle East, North America, and Europe (Stukenbrock et al. 2007).\n"
      ],
      "metadata": {
        "id": "0yoOMkQ-Gv3m"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget \"https://raw.githubusercontent.com/emgoss/PLP6621C/refs/heads/main/HPPD.fasta\" -O HPPD.fasta"
      ],
      "metadata": {
        "id": "678WlOthR1OE"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Set an environmental variable for the input file so we can don't have to give the full path for each command. You can check the correct path of the HPPD.fasta file by opening the Files directory on the left size of the screen (folder icon), right click on the file name, copy path, and paste into the below code after \"=\"."
      ],
      "metadata": {
        "id": "oOcvw9siOt2K"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%env INPUT_FILE=/content/HPPD.fasta"
      ],
      "metadata": {
        "id": "TMv-LjSXOMPJ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**4. Run ModelFinder in IQTREE on the HPPD.fasta alignment using the -m TESTONLY command.**\n",
        "\n",
        "TESTONLY is telling IQTREE to find a model only and not run a phylogenetic analysis, which we'll do later.\n"
      ],
      "metadata": {
        "id": "G2onM2OhJcCp"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!/usr/bin/iqtree2 -s $INPUT_FILE -st DNA -m TESTONLY --prefix HPPD_test -nt AUTO"
      ],
      "metadata": {
        "id": "LwQ6KcdO4AP2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "In the output, you can see the names of all the sequences and the results of the analysis. See the bottom of the table of models for the best model according to three criteria that take into account the likelihood of the model for these genetic data and the number of parameters in the model.\n",
        "\n",
        "**5. In the below box, provide the model chosen using Corrected Akaike Information Criterion (AICc) and Bayesian Information Criterion (BIC).**"
      ],
      "metadata": {
        "id": "zc9BnFUqLHKw"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "***Your answer here:***\n",
        "\n",
        "Corrected Akaike Information Criterion:\n",
        "\n",
        "Bayesian Information Criterion:\n",
        "\n",
        "---\n",
        "\n"
      ],
      "metadata": {
        "id": "K9mHwmQjG4Ng"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "See [this page](https://iqtree.github.io/doc/Substitution-Models#dna-models) for basic descriptions of the modelfinder abbreviations. Scroll to the bottom to find the add-on abbreviations.\n",
        "\n",
        "**6. What is the name of the models that were chosen?**\n",
        "\n",
        "***Your answer here:***\n",
        "\n",
        "---\n",
        "\n",
        "**7. What does this mean in terms of base frequencies and DNA substitution rates? Explain substitution rates in terms of transitions versus transversions and purines versus pyrimidines.**\n",
        "\n",
        "***Your answer here:***\n",
        "\n",
        "\n",
        "---"
      ],
      "metadata": {
        "id": "iYHGKY0sFpQA"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**8. Now try the new ModelFinder using -m TESTONLYNEW.**"
      ],
      "metadata": {
        "id": "b3kFtxf-K8hT"
      }
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "cb33baf0"
      },
      "source": [
        "!/usr/bin/iqtree2 -s $INPUT_FILE -st DNA -m TESTONLYNEW --prefix HPPD_testnew -nt AUTO"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "**9. What is the best model using the new ModelFinder?**\n",
        "\n",
        "***Your answer here:***\n",
        "\n",
        "New model Corrected Akaike Information Criterion:\n",
        "\n",
        "New model Bayesian Information Criterion:\n"
      ],
      "metadata": {
        "id": "LQGDyYQa6ARS"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**10. Now we will infer phylogenetic trees using the BIC-selected model.**\n",
        "\n",
        "You will need to paste your selected model (use BIC, this is common in the literature) into the code below (remove the brackets [])."
      ],
      "metadata": {
        "id": "pg9Baz6mOHwi"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!/usr/bin/iqtree2 -s $INPUT_FILE -m [paste model here] -nt AUTO\n",
        "\n",
        "# e.g. !/usr/bin/iqtree2 -s $INPUT_FILE -m JC -nt AUTO"
      ],
      "metadata": {
        "id": "I1n0DeFbnxly"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**11. What were the estimated DNA substitution rates used to infer the tree? Hint: They are relative to A-C, which is set to 1.0.**\n",
        "\n",
        "***Your answer here:***\n",
        "\n",
        "---"
      ],
      "metadata": {
        "id": "f023eXrjHztP"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**12. Now run the tree again but with a non-optimal model for the tree.**\n",
        "\n",
        "The GTR model has the most free parameters of the substitution models. We will add variation among sites across the alignment by adding four categories of  rate variation (G4)."
      ],
      "metadata": {
        "id": "1m21SKE-JYev"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!/usr/bin/iqtree2 -s $INPUT_FILE -m GTR+G4 --prefix HPPD_GTRGtree -nt AUTO"
      ],
      "metadata": {
        "id": "GeIjRtk0KFep"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**13. Take a look at your trees.**\n",
        "\n",
        "Install biopython and use the phylo package to load and view the tree. This is a quick and dirty ASCII version.\n",
        "\n",
        "Or download the treefile and import into your favorite treeviewer. One option that doesn't require installing software is https://itol.embl.de/"
      ],
      "metadata": {
        "id": "Yo-dWKcYp5mW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install biopython\n"
      ],
      "metadata": {
        "id": "PR99jLcLoipX"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio import Phylo\n",
        "# original tree\n",
        "tree = Phylo.read(\"/content/HPPD_testnew.treefile\", \"newick\") #check path\n",
        "Phylo.draw_ascii(tree)"
      ],
      "metadata": {
        "id": "WXmWgD2JqnxX"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**14. The isolates with \"S\" prefix are from wild relatives of wheat in Iran. All other isolates were collected from wheat in Iran (MG), Israel (Is), US (Or), and Europe (Ch). Based on your tree, what can you say about the \"S\" isolates?**\n",
        "\n",
        "***Your answer here:***\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "**15. Compare the tree inferred with the optimal model to the tree using the sub-optimal model. Does the tree change when you use a sub-optimal model?**\n",
        "\n",
        "***Your answer here:***\n",
        "\n",
        "---"
      ],
      "metadata": {
        "id": "VbDWvsUqqyw4"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "References cited:\n",
        "\n",
        "S. Kalyaanamoorthy, B.Q. Minh, T.K.F. Wong, A. von Haeseler, and L.S. Jermiin (2017) ModelFinder: fast model selection for accurate phylogenetic estimates. Nat. Methods, 14:587–589. DOI: 10.1038/nmeth.4285\n",
        "\n",
        "Stukenbrock et al. (2017) Mol. Biol. Evol. 24(2):398–411. https://doi.org/10.1093/molbev/msl169"
      ],
      "metadata": {
        "id": "8Eszm9sz95X3"
      }
    }
  ]
}