{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyM+nB8yYI2oQ87NJ767O96A",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/emgoss/PLP6621C/blob/main/Module12.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "## Module 12 Assignment: Detecting selection on coding sequences\n",
        "\n",
        "You will run tests of selection using HyPhy software and interpret the results. HyPhy runs on the command line and requires specific input files and formats. If you don't want more help than the command line provides and you're willing to wait for your job in the queue, you can use the web interface: https://www.datamonkey.org/.\n",
        "\n",
        "There is a lot of useful information on the web about HyPhy. Scroll down on this website to see a brief overview of tests available in HyPhy: https://hyphy.org/getting-started/"
      ],
      "metadata": {
        "id": "Jwm6ZBuWncje"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**1. Install HyPhy.**\n",
        "\n",
        "Remember, each time you use it on Colab, you will have to re-install.\n",
        "\n",
        "First, obtain the software from GitHub."
      ],
      "metadata": {
        "id": "jx6ws9ysqSzd"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!git clone https://github.com/veg/hyphy.git"
      ],
      "metadata": {
        "id": "wbePc_7yPF6Z"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Go to the hyphy directory before executing the install commands."
      ],
      "metadata": {
        "id": "NW-7wtT2qsvp"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "cd hyphy"
      ],
      "metadata": {
        "id": "L0_iepzEPbqK"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Compile and install the software."
      ],
      "metadata": {
        "id": "xHcaN4eerJhy"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!cmake .\n",
        "!make install"
      ],
      "metadata": {
        "id": "Lm2dxo_nPftO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Make sure it works by running the help option."
      ],
      "metadata": {
        "id": "Ibn3h6fMrSD3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!hyphy --help"
      ],
      "metadata": {
        "id": "HTElK1CWTy8t"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "\n",
        "**2. Dataset preparation**\n",
        "\n",
        "We will run data from ORF 4 of the coat protein coding gene of the rice yellow mottle virus from an arbitrary subset of the WA261 dataset in this paper:\n",
        "https://academic.oup.com/ve/article/7/2/veab072/6353248\n",
        "\n",
        "Datamonkey.org will automatically remove stop codons from the end of an alignment, but the command line HyPhy does not. After downloading and aligning the sequences from GenBank, I deleted the stop codon from the end of the alignment.\n",
        "\n",
        "Download the [fasta alignment](https://ufl.instructure.com/courses/540315/files/103058452?wrap=1) and the [newick tree file](https://ufl.instructure.com/courses/540315/files/103058453?wrap=1) from the Module 12 Assignment.\n",
        "\n",
        "---"
      ],
      "metadata": {
        "id": "QyBaDUU8W0Hr"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**3. Check for recombination.**\n",
        "\n",
        "The first analysis we will do is GARD to check for recombination in our data. All of the following analyses should be done on non-recombinant sequences. GARD will identify recombination and associated breakpoints to divide the alignment into non-recombinant partitions.\n",
        "\n",
        "The input required for GARD, and for most of the HyPhy analyses, is the in-frame coding sequence alignment and an associated newick tree. You can run help to check this."
      ],
      "metadata": {
        "id": "DZWFmv-1tIY9"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!hyphy gard --help"
      ],
      "metadata": {
        "id": "DyiD0Drhx2Lp"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "In this case, I inferred the tree using RAxML for you. (If you use DataMonkey, it will generate a tree for you.)\n",
        "\n",
        "I was encountering a numerical error in the 2-breakpoint analysis and HyPhy told me to deal with this using ENV=TOLERATE_NUMERICAL_ERRORS=1."
      ],
      "metadata": {
        "id": "d-yHVInMx75T"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!hyphy gard --alignment rice_yellow_mottle_no_stop.fasta --tree rice_yellow_mottle_RAxML.newick ENV=TOLERATE_NUMERICAL_ERRORS=1;"
      ],
      "metadata": {
        "id": "s7ggFH63HzeW"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Check out the GARD output and the files it generated. Which model has the lowest AICc?\n",
        "\n",
        "We will use the GARD output for the following analyses."
      ],
      "metadata": {
        "id": "SoK3l9o8z3XJ"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**4. Test for pervasive selection on individual codons using FEL.**\n",
        "\n",
        "Scroll down https://hyphy.org/getting-started/ and https://hyphy.org/methods/selection-methods/ to learn about FEL.\n",
        "\n"
      ],
      "metadata": {
        "id": "S6InjwGQ0ou_"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!hyphy fel --alignment rice_yellow_mottle_no_stop.fasta.best-gard"
      ],
      "metadata": {
        "id": "7RcQNPZ8U5tp"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**5. Compare to output from SLAC.**\n"
      ],
      "metadata": {
        "id": "gZT1truK1mzL"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!hyphy slac --alignment rice_yellow_mottle_no_stop.fasta.best-gard"
      ],
      "metadata": {
        "id": "mAmetknqQFSF"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**6. Compare to output from FUBAR.**\n"
      ],
      "metadata": {
        "id": "J-V0NCS22Dbo"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!hyphy fubar --alignment rice_yellow_mottle_no_stop.fasta.best-gard"
      ],
      "metadata": {
        "id": "7E2PqqQqOwv5"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "**Question 1**\n",
        "\n",
        "How many sites have experienced pervasive positive selection using each method?\n",
        "\n",
        "Positively selected sites for:\n",
        "*   FEL\n",
        "*   SLAC\n",
        "*   FUBAR\n",
        "\n",
        "\n",
        "---"
      ],
      "metadata": {
        "id": "dbofU-dH2sum"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**7. Run MEME.**\n",
        "\n",
        "The final analysis is MEME. How is MEME different from the previous selection analyses?"
      ],
      "metadata": {
        "id": "hrm9APWZ6PSp"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!hyphy meme --alignment rice_yellow_mottle_no_stop.fasta.best-gard"
      ],
      "metadata": {
        "id": "Ujpu6QoMQH4q"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "---"
      ],
      "metadata": {
        "id": "uss7N5PQVWmW"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "**Question 2**\n",
        "\n",
        "A. What is MEME testing that the three other methods are not?\n",
        "\n",
        "B. How is this difference reflected in the output?\n",
        "\n",
        "\n",
        "---"
      ],
      "metadata": {
        "id": "yz322HbGLPqZ"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Question 3**\n",
        "\n",
        "A. Which methods use a maximum likelihood framework and which use Bayesian inference?\n",
        "*   FEL - ML or Bayesian\n",
        "*   SLAC\n",
        "*   FUBAR\n",
        "*   MEME\n",
        "\n",
        "B. What is the criterion used to determine if a site has statistically significant selection for each method?\n",
        "*   FEL\n",
        "*   SLAC\n",
        "*   FUBAR\n",
        "*   MEME\n",
        "\n",
        "---"
      ],
      "metadata": {
        "id": "w9v9CZDTICqo"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Question 4**\n",
        "\n",
        "These sequences are from isolates collected from West Africa over decades.\n",
        "\n",
        "A. What specifically do the results of these tests tell you about the evolution of the coat protein of the rice yellow mottle virus?\n",
        "\n",
        "B. Based on the your results and the biology of the virus, propose two testable hypotheses about the evolution of this virus that follow from the results. You can use the paper that these data came from to help you, but your hypotheses need use your new knowledge regarding selection on the coat protein.\n",
        "\n",
        "---"
      ],
      "metadata": {
        "id": "pktn9RQaP1wx"
      }
    }
  ]
}