{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyPrF3zpg2MWN3eqdQF1CY3m",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "ir",
      "display_name": "R"
    },
    "language_info": {
      "name": "R"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/emgoss/PLP6621C/blob/main/Module5_R.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Module 5 Assignment: Recombination\n",
        "\n",
        "**Purpose:**\n",
        "The purpose of this assignment is to learn how to calculate and interpret the index of association for microsatellite and SNP data sets. This statistic has been used for both bacterial and fungal populations to test for significant linkage disequilibrium. This assignment contributes to overall course objectives of discussing how analysis of population genetic data can be used to learn about microbial populations, evaluating the assumptions, limitations, and appropriate use of population genetic analysis in the plant pathology literature, and conducting conduct population genetic analyses using existing data sets.\n",
        "\n",
        "\n",
        "We will use R packages for this assignment. The index of association for SSRs analysis is taken verbatum from this tutorial, so please also refer to the source material:\n",
        "https://grunwaldlab.github.io/Population_Genetics_in_R/index.html\n",
        "\n",
        "For the SNP data, I have also used this:\n",
        "https://grunwaldlab.github.io/poppr/articles/mlg.html\n",
        "and Step5_Data_processing.Rmd from here:\n",
        "https://github.com/grunwaldlab/rubi-gbs\n"
      ],
      "metadata": {
        "id": "t6H-EgSdhnVN"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Calculating the Index of Association with Microsatellite Data\n",
        "\n",
        "**1. Install the first R package we need, [poppr](https://grunwaldlab.github.io/poppr/reference/poppr-package.html).**"
      ],
      "metadata": {
        "id": "zcQ26s7-11pf"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "lTff09BbhX8d"
      },
      "outputs": [],
      "source": [
        "install.packages(\"poppr\")\n",
        "library(\"poppr\")"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**2. Load the data**\n",
        "\n",
        "The data used in this example, called monpop, are from Everhart & Scherm, 2015. This is a microsatellite data set  of 13 loci for 694 individuals of the haploid fungal pathogen *Monilinia fructicola* that infects peach flowers and fruits in commercial orchards. The data came from four trees within a single orchard (trees 26, 45, and 7). Each tree was sampled in 2009, 2010, and/or 2011. Additionally, each sample was noted as to whether it came from a blossom or a fruit. This example data set is included with the poppr package.\n",
        "\n",
        "Poppr allows setting the hierachy of your data so that you can easily analyze different population or grouping levels. These are called \"strata\".\n",
        "\n",
        "The steps for working with stratified data include:\n",
        "\n",
        "1. Import data set with samples labeled according to strata\n",
        "2. Define the stratifications for the data\n",
        "3. Setting the stratification(s) that you want to have analyzed\n",
        "\n",
        "The easiest way to work with stratified data is to label each sample using an underscore “_” to separate each level. This was already done for the monpop data, where each sample was coded hierarchically by tree, year, and symptom in the following format: “Tree_Year_Symptom”.\n",
        "\n",
        "The data we will use is distributed with the poppr package so we don't need to import it, just load it.\n"
      ],
      "metadata": {
        "id": "GKaKwfpmo36i"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "data(monpop)\n",
        "monpop #view a summary of the data"
      ],
      "metadata": {
        "id": "gnesn_AWoO_2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "The **genclone object** is a type of container that holds the data. In this case, it structured for small genotype data sets that will be clone corrected.\n",
        "\n",
        "**Genotype information** shows us that the data contains 264 multilocus genotypes among 694 haploid individuals with 13 loci.\n",
        "\n",
        "**Population information** has two items, the stratifications and the populations defined. You can think of stratifications as the index names for each of the hierarchical levels within your data (so for our data it should be Tree, Year, and Symptom). By default, however, no stratifications are defined and so this is “Pop”, which is the entire dataset of 694 individuals. Because we labeled each sample according to stratification, populations defined shows us our data has 12 groups defined: 7_09_BB, 26_09_BB, 26_09_FR, 7_09_FR, 26_10_BB, 45_10_BB, 79_10_BB, 79_10_FR, 26_10_FR, 45_10_FR, 26_11_BB, and 26_11_FR.\n",
        "\n"
      ],
      "metadata": {
        "id": "7roZeAhksstv"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**3. Assign stratifications**\n",
        "\n",
        "We imported the data that has three stratifications “Tree_Year_Symptom”. In order to analyze our data according to any combination of those three stratifications, we need to tell poppr that the 12 groups should be split by tree, year, and/or symptom. Thus, the first step is to split our data according to strata so that we can access each of the three hierarchical levels in the data. The splitStrata command is used to index the three stratifications:"
      ],
      "metadata": {
        "id": "YerLW9w2tRRF"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "splitStrata(monpop) <- ~Tree/Year/Symptom\n",
        "monpop # view the data summary again"
      ],
      "metadata": {
        "id": "wrl0CD3WpUjp"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "What changed?\n",
        "\n",
        "After splitting the data, there are 3 strata: “Tree Year Symptom”."
      ],
      "metadata": {
        "id": "4cGPjY31t1fS"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**4. Setting different strata (populations)**\n",
        "\n",
        "To analyze the data according to Symptom:"
      ],
      "metadata": {
        "id": "3BcjnHikuUmr"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "setPop(monpop) <- ~Symptom\n",
        "monpop # check the populations"
      ],
      "metadata": {
        "id": "KBCmgoJxueTB"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Now there are two populations, blossom blight (BB) and fruit rot (FR).\n",
        "\n",
        "If we wanted to define the symptoms according to tree, we would use the following:"
      ],
      "metadata": {
        "id": "_wtjo-2tjSpn"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "setPop(monpop) <- ~Symptom/Tree\n",
        "monpop"
      ],
      "metadata": {
        "id": "HxQkWZFPu15t"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Note that these populations combine the three different years."
      ],
      "metadata": {
        "id": "DgHbcBmGvBou"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**5. Clone correction**\n",
        "\n",
        "When dealing with clonal populations, analyses are typically conducted with and without clone correction. Clone correction is a method of censoring a data set such that only one individual per MLG is represented per population (Milgroom, 1996; Grünwald et al., 2003; Grünwald & Hoheisel, 2006). This technique is commonly used with the index of association and genotypic diversity measures since clone corrected populations approximate behavior of sexual populations. Since we want to only observe unique genotypes per population, clone correction requires specification of the stratifications (population) at which clones should be censored. This section will show how to clone correct at a specific stratification and also compare the results with uncorrected data."
      ],
      "metadata": {
        "id": "x6WfVRAdvOLn"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Looking at the above summaries of the monpop data set, we see that there are only 264 genotypes in the total sample of 694 individuals. This indicates that there are many individuals that share identical multilocus genotypes.\n",
        "\n",
        "We then have to decide how we want to clone correct, which must be informed by the biology of the organism and sampling strategy. Take a break from the code and **read through the sampling methods** https://apsjournals.apsnet.org/doi/pdf/10.1094/PHYTO-03-14-0088-R. What is the most biologically appropriate way to clone-correct these data?\n"
      ],
      "metadata": {
        "id": "Xul51PjWv1xM"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "We will try clone correcting using different population definitions and compare the results. To set our populations to tree and year, we will use the “Tree/Year” strata then clone correct, assigning the clone corrected data to a new variable so that we don't write over the original data set."
      ],
      "metadata": {
        "id": "3Sp6ym4KdfcZ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "setPop(monpop) <- ~Tree/Year\n",
        "monpop # check that populations are defined as you wanted"
      ],
      "metadata": {
        "id": "JImSA_3HySav"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "mcc_TY <- clonecorrect(monpop, strata = ~Tree/Year, keep = 1:2) # clone correct by both tree and year and assign to mcc_TY\n",
        "mcc_TY # view the new data set"
      ],
      "metadata": {
        "id": "r7qc3GFxzKmI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Now there are only 278 individuals because only one individual of each genotype has been retained per tree in each year.\n",
        "\n",
        "Now clone correct by year and symptom.\n"
      ],
      "metadata": {
        "id": "Ir8EzVA-zY2I"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "setPop(monpop) <- ~Year/Symptom\n",
        "monpop # check that populations are defined as you wanted"
      ],
      "metadata": {
        "id": "j1-AfJkg3Op_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "mcc_YS <- clonecorrect(monpop, strata = ~Year/Symptom, keep=1:2) # clone correct by both year and symptom and assign to mcc_TY\n",
        "mcc_YS # view the new data set"
      ],
      "metadata": {
        "id": "8Vwsyx0j3ejv"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Question 1:**\n",
        "Compare results of clone correction using the above different population definitions (strata). Explain why the number of individuals differ among different population definitions.\n",
        "\n"
      ],
      "metadata": {
        "id": "soBDLxKidpCt"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**6. Index of association.**\n",
        "\n",
        "Calculate the index of association (Ia) for the 9_BB population. First we must pull the population out from the monpop dataset, then calculate the Ia. (We have to put a letter in front of the population name to not get an error in R.)\n",
        "\n",
        "The actual value of Ia will be one of 1000 calculations of Ia. The rest use 999 random permutations of the data among individuals. This will give a minimum p-value of 0.001. If we want to calculate to more decimal places, we need to add an order of magnitude more random samples.\n",
        "\n",
        "The output includes the value of Ia and associated p-value from the permutations (p.Ia), and rbarD and associated p-value (p.rD). Remember that rbarD takes similar values across different numbers of loci whereas the magnitude of Ia depends on the number of loci genotyped."
      ],
      "metadata": {
        "id": "qWHE05mGEHkg"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "T9_BB <- popsub(monpop, \"9_BB\")\n",
        "ia(T9_BB, sample = 999)"
      ],
      "metadata": {
        "id": "njR5dvUj_A8m"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Now do the same for the *clone corrected* 9_BB population."
      ],
      "metadata": {
        "id": "GwgLkEaHFFdI"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "T9_BBcc <- popsub(mcc_YS, \"9_BB\")\n",
        "ia(T9_BBcc, sample = 999)"
      ],
      "metadata": {
        "id": "evF1eRCb_eHl"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "The values of Ia and rbarD are lower for the clone corrected sample, but linkage equilibrium is rejected for both.\n",
        "\n",
        "You can also quickly obtain the values for Ia and rbarD using the command \"poppr\". But you can't get p-values this way. We will talk about the other values in this table in Module 7."
      ],
      "metadata": {
        "id": "IOTessSuFKUE"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "poppr(monpop)"
      ],
      "metadata": {
        "id": "FxWN-Day_08k"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "poppr(mcc_YS)"
      ],
      "metadata": {
        "id": "-q90lHZ_iP68"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "The lowest values of Ia and rbarD were obtained for 11_BB. Note that for this populations, N = MLG meaning that each sampled isolate had a different genotype. Therefore, the values for the clone corrected 11_BB is the same as the original 11_BB. Let's see if we have significant linkage disequilibrium in this population."
      ],
      "metadata": {
        "id": "ESEkpXH9vpe3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "T11_BB <- popsub(monpop, \"11_BB\")\n",
        "ia(T11_BB, sample = 999)"
      ],
      "metadata": {
        "id": "iXA6MB4-AefH"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "There is not significant linkage disequilibrium in the 11_BB population."
      ],
      "metadata": {
        "id": "avtfVRO0wgSm"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Question 2:**\n",
        "Practice calculating the index of association. Clone correct by *year*, calculate the index of association for each year including the significance test. Summarize the results in a table. What is your interpretation of these results?\n",
        "\n",
        "**Question 3:**\n",
        "What biological explanations could explain the observed results? You may use the Everhart and Scherm paper for help."
      ],
      "metadata": {
        "id": "h7UPdJrAw0FR"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Index of association using SNP data"
      ],
      "metadata": {
        "id": "KFWeRlD8jE6_"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "For SNP data, we need to take another approach to clone correction, which is to identify multilocus lineages rather than multilocus genotypes.\n",
        "\n",
        "We will use a data set of 94 samples of the red raspberry pathogen *Phytophthora rubi* (Tabima et al., 2018). This pathogen is diploid oomycete. Populations were obtained by sampling individual pathogen strains from roots of infected red raspberry in the states of California (CA), Oregon (OR), and Washington (WA). Currently, there is little information about the population structure of *P. rubi* in the western USA.\n",
        "\n",
        "Note that the Phytophthora rubi files we're using are from the Population Genetics in R tutorial, not the exact analysis in the paper.\n",
        "\n",
        "To obtain variant calls in form of VCF data, the FASTQ reads from Illumina sequencing were mapped to the reference genome of *P. rubi* (Tabima et al., 2018) using bowtie2 (Langmead & Salzberg, 2012). Variants were called using the GATK HaplotypeCaller (McKenna et al., 2010). This data was further filtered in vcfR using read depths (DP) and mapping qualities (MQ).\n",
        "\n",
        "**7. Install packages for importing the data and graphing the results.**"
      ],
      "metadata": {
        "id": "0nOHsog2jjr-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "install.packages(c(\"vcfR\",\"RColorBrewer\",\"ggplot2\"))\n",
        "library(vcfR)\n",
        "library(RColorBrewer)\n",
        "library(ggplot2)"
      ],
      "metadata": {
        "id": "_PrD4ik7jEUh"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**8. First import the dataset into your Colab.**"
      ],
      "metadata": {
        "id": "OEyXhNkmqUxr"
      }
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "a63c3895"
      },
      "source": [
        "# Define the URL of file\n",
        "data_url <- \"https://grunwaldlab.github.io/Population_Genetics_in_R/prubi_gbs.vcf.gz\"\n",
        "\n",
        "# Define the destination path and filename\n",
        "data_path <- \"./prubi_gbs.vcf.gz\"\n",
        "\n",
        "# Download in binary mode\n",
        "download.file(url = data_url, destfile = data_path, mode = \"wb\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**9. Read SNP data from the compressed VCF file (we will talk about these in Module 6) using package vcfR.**"
      ],
      "metadata": {
        "id": "iDNq503J5U6k"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "rubi.vcf <- read.vcfR(\"prubi_gbs.vcf.gz\")\n",
        "rubi.vcf"
      ],
      "metadata": {
        "id": "SO3J_ywpjBDn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "The file population_data.gbs.txt is a tab-delimited text file that includes the name of the sample, country of origin, and the population from where it was sampled. The file is available for download at: population_data.gbs.txt. This link will likely open the data in a browser. Save the data onto your hard-drive as an ASCII text file with the same name.\n",
        "\n",
        "**10. Now download and import the metadata (population assignments).**"
      ],
      "metadata": {
        "id": "lrCNHb6DrhHF"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "file_url <- \"https://grunwaldlab.github.io/Population_Genetics_in_R/population_data.gbs.txt\"\n",
        "dest_path <- \"./population_data.gbs.txt\"\n",
        "download.file(url = file_url, destfile = dest_path, method = \"auto\")"
      ],
      "metadata": {
        "id": "lAvLLPddBCwy"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pop.data <- read.table(\"population_data.gbs.txt\", sep = \"\\t\", header = TRUE)\n",
        "pop.data"
      ],
      "metadata": {
        "id": "VWheB0BPr3xH"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**11. Convert data format for analysis.**\n",
        "\n",
        "For the analysis, we need to convert the vcfR object into a genlight object (these are different data structures), assign the ploidy, and associate the population assignments."
      ],
      "metadata": {
        "id": "1NEVgDMesB9s"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "gl.rubi <- vcfR2genlight(rubi.vcf)\n",
        "ploidy(gl.rubi) <- 2\n",
        "pop(gl.rubi) <- pop.data$State\n",
        "gl.rubi"
      ],
      "metadata": {
        "id": "fEEJ-dT4Bu6C"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**12. Index of association using samp.ia**\n",
        "\n",
        "For SNPs, we don't necessarily want to use all loci for calculating the index of association, because it's overkill. The samp.ia function randomly samples sites to calculate the index of association. The default is to sample 100 SNPs and do this 100 times."
      ],
      "metadata": {
        "id": "DC8MTsf26tFY"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "rubi.ia <- samp.ia(gl.rubi)\n",
        "rubi.ia"
      ],
      "metadata": {
        "id": "qYEWgXzM78Mo"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Because that output isn't particularly useful as is, we need to summarize it and better yet compare it to data with known structures. The following code generates data with the same number of individuals and SNPs with varying numbers of linked sites to mimic linkage disequilibrium. The actual and simulated data sets are then compared in a bar plot."
      ],
      "metadata": {
        "id": "J_stsOii8B4o"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Simulated populations\n",
        "### No structure (admixed pops)\n",
        "sex <- glSim(94, 608, ploid=2, LD=T)\n",
        "### Structure (clonal pops)\n",
        "clone <- glSim(94, 608, n.snp.struc=608, ploid=2, LD = T)\n",
        "### Semi-clonal - half clonal\n",
        "semi_clone <- glSim(94, 608, n.snp.struc=304, ploid=2, LD=T)\n",
        "### Most-clonal - 2/3 clonal\n",
        "most_clone <- glSim(94, 608, n.snp.struc=405, ploid=2, LD=T)\n",
        "\n",
        "## IA sex\n",
        "ia.sex <- samp.ia(sex,quiet = T)\n",
        "## IA clone\n",
        "ia.clone <- samp.ia(clone, quiet = T)\n",
        "## IA.semiclone\n",
        "ia.semi <- samp.ia(semi_clone, quiet = T)\n",
        "## IA.mostclone\n",
        "ia.most <- samp.ia(most_clone, quiet = T)\n",
        "\n",
        "# Summarizing data frames\n",
        "d1 <- data.frame(rubi.ia, rep(\"dataset\", length(rubi.ia)))\n",
        "d2 <- data.frame(ia.sex, rep(\"sexual\", length(ia.sex)))\n",
        "d3 <- data.frame(ia.clone, rep(\"clone\", length(ia.clone)))\n",
        "d4 <- data.frame(ia.semi, rep(\"semi-clone\", length(ia.semi)))\n",
        "d5 <- data.frame(ia.most, rep(\"most-clone\", length(ia.semi)))\n",
        "colnames(d1) <- c(\"ia\",\"dset\")\n",
        "colnames(d2) <- c(\"ia\",\"dset\")\n",
        "colnames(d3) <- c(\"ia\",\"dset\")\n",
        "colnames(d4) <- c(\"ia\",\"dset\")\n",
        "colnames(d5) <- c(\"ia\",\"dset\")\n",
        "ia.total <- rbind(d1, d2, d3, d4, d5)\n",
        "\n",
        "# Plot\n",
        "ggplot(ia.total,aes(dset,ia,fill=dset)) + geom_boxplot() + xlab(\"Dataset\") + ylab(\"Index of association\")\n"
      ],
      "metadata": {
        "id": "LvixrGmSD4tm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**13. Clone correcting SNP datasets.**\n",
        "\n",
        "For SNP datasets, we need to take a different approach to clone correction. We need to define at what genetic distance to clone correct.\n",
        "\n",
        "The first step is to move the data into a snpclone object, which can handle clone correction."
      ],
      "metadata": {
        "id": "TCaavsPL9IWE"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "sc.rubi <- as.snpclone(gl.rubi)\n",
        "sc.rubi"
      ],
      "metadata": {
        "id": "mBOdnoKcDDT6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Here is the number of multilocus genotypes in the dataset."
      ],
      "metadata": {
        "id": "q6ZyYJmn_gi0"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "nmll(sc.rubi)"
      ],
      "metadata": {
        "id": "yDa08Z3K-p27"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Now we will caculate the genetic distance between all individuals in the sc.rubi data set and use the cutoff_predictor function to tell us which genetic distances we should consider using to define our multilocus lineages."
      ],
      "metadata": {
        "id": "9FLXAAxT-xOb"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "rubi_filtered <- filter_stats(sc.rubi, distance = bitwise.dist, plot = TRUE)\n",
        "print(farthest_thresh <- cutoff_predictor(rubi_filtered$farthest$THRESHOLDS))\n",
        "print(average_thresh  <- cutoff_predictor(rubi_filtered$average$THRESHOLDS))\n",
        "print(nearest_thresh  <- cutoff_predictor(rubi_filtered$nearest$THRESHOLDS))\n",
        "\n"
      ],
      "metadata": {
        "id": "1C6ouc51u5af"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "The resulting histogram of pairwise genetic distances is bimodal. This suggests that there are individuals that are more genetically close than we would expect based on pairwise genetic distances across the dataset (the distribution on the right).\n",
        "\n",
        "We will use the \"farthest_thresh\" cutoff as that seems the most likely to capture the potentially clonal isolates in that left node in the histogram."
      ],
      "metadata": {
        "id": "CvfrcGlv_qFH"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "mlg.filter(sc.rubi, distance = bitwise.dist, algorithm = \"f\") <- farthest_thresh\n",
        "sc.rubi"
      ],
      "metadata": {
        "id": "H1iZjcopDkvv"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Now we have 53 multilocus lineages among our 94 isolates."
      ],
      "metadata": {
        "id": "DW_-rigJAVgD"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "nmll(sc.rubi)"
      ],
      "metadata": {
        "id": "7uyKL0Zu_VD8"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Question 4:** What are your inferences about linkage disequilibrium and sexual reproduction in the *P. rubi* population based on this analysis?"
      ],
      "metadata": {
        "id": "-xyBqWq-DEu5"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Question 5:** See assignment for a graph, which shows data for populations of a pathogen on a widely planted crop. The dark diamonds indicate populations in a putative source region and the light diamonds represent populations in other parts of the world. The data shown is: rD, index of association after clone correction plotted against CG:N, which is the number of different clonal groups (CG, i.e., multilocus lineages) scaled by sample size (N). What is figure showing about these populations?"
      ],
      "metadata": {
        "id": "Bl_BKjTpEKII"
      }
    }
  ]
}